op-interop-mon/metrics/metrics.go

```diff
@@ -0,0 +1,160 @@
+package metrics
+
+import (
+	"fmt"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	opmetrics "github.com/ethereum-optimism/optimism/op-service/metrics"
+)
+
+const Namespace = "op_interop_mon"
+
+var _ opmetrics.RegistryMetricer = (*Metrics)(nil)
+
+type Metricer interface {
+	RecordInfo(version string)
+	RecordUp()
+	RecordInitiatingMessageStats(chainID string, blockHeight uint64, status string, value float64)
+	RecordExecutingMessageStats(chainID string, blockHeight uint64, blockHash string, status string, value float64)
+	RecordTerminalStatusChange(executingChainID string, initiatingChainID string, value float64)
+
+	opmetrics.RefMetricer
+	opmetrics.RPCMetricer
+}
+
+type Metrics struct {
+	ns       string
+	registry *prometheus.Registry
+	factory  opmetrics.Factory
+
+	opmetrics.RefMetrics
+	opmetrics.RPCMetrics
+
+	info prometheus.GaugeVec
+	up   prometheus.Gauge
+
+	// Message metrics
+	executingMessages     prometheus.GaugeVec
+	initiatingMessages    prometheus.GaugeVec
+	terminalStatusChanges prometheus.GaugeVec
+}
+
+var _ Metricer = (*Metrics)(nil)
+
+func NewMetrics(procName string) *Metrics {
+	if procName == "" {
+		procName = "default"
+	}
+	ns := Namespace + "_" + procName
+
+	registry := opmetrics.NewRegistry()
+	factory := opmetrics.With(registry)
+
+	return &Metrics{
+		ns:       ns,
+		registry: registry,
+		factory:  factory,
+
+		RefMetrics: opmetrics.MakeRefMetrics(ns, factory),
+		RPCMetrics: opmetrics.MakeRPCMetrics(ns, factory),
+
+		info: *factory.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: ns,
+			Name:      "info",
+			Help:      "Information about the monitor",
+		}, []string{
+			"version",
+		}),
+		up: factory.NewGauge(prometheus.GaugeOpts{
+			Namespace: ns,
+			Name:      "up",
+			Help:      "1 if the op-interop-mon has finished starting up",
+		}),
+		executingMessages: *factory.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: ns,
+			Name:      "executing_messages",
+			Help:      "Number of messages being executed",
+		}, []string{
+			"chain_id",
+			"block_height",
+			"block_hash",
+			"status",
+		}),
+		initiatingMessages: *factory.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: ns,
+			Name:      "initiating_messages",
+			Help:      "Number of messages being initiated",
+		}, []string{
+			"chain_id",
+			"block_height",
+			"status",
+		}),
```

Comment:
I want to make sure I have my mental model correct. 

Am I right in thinking the entrypoint to discover initiating messages is actually just the executing messages, which we get from the `CrossL2Inbox` on the receiving chain?

In other words, any log _could_ be an initiating message, and we don't really know until an executing message references it.

Reply:
Yes that is right. The Help string could be massively improved because technically *every* log message is an Initiating Message. But like you say, we don't know which ones we care about until they are referenced.

A better way of describing it is "Number of Initiating Messages being referenced by Executing Messages" (organized by chain, block, and status).